apps/src/pythonlab/PythonEditor.tsx

```diff
@@ -3,37 +3,61 @@ import {darkMode} from '@cdo/apps/lab2/views/components/editor/editorThemes';
 import {python} from '@codemirror/lang-python';
 import moduleStyles from './python-editor.module.scss';
 import {useDispatch} from 'react-redux';
-import {appendOutput, resetOutput, setCode} from './pythonlabRedux';
+import {appendOutput, resetOutput, setSource} from './pythonlabRedux';
 import Button from '@cdo/apps/templates/Button';
 import {runPythonCode} from './pyodideRunner';
 import {useFetch} from '@cdo/apps/util/useFetch';
 import CodeEditor from '@cdo/apps/lab2/views/components/editor/CodeEditor';
 import {useAppSelector} from '@cdo/apps/util/reduxHooks';
+import Lab2Registry from '@cdo/apps/lab2/Lab2Registry';
 
 interface PermissionResponse {
   permissions: string[];
 }
 
 const PythonEditor: React.FunctionComponent = () => {
-  const code = useAppSelector(state => state.pythonlab.code);
+  const source = useAppSelector(state => state.pythonlab.source);
   const codeOutput = useAppSelector(state => state.pythonlab.output);
   const {loading, data} = useFetch('/api/v1/users/current/permissions');
   const dispatch = useDispatch();
-
-  const onCodeUpdate = (code: string) => dispatch(setCode(code));
   const editorExtensions = [python(), darkMode];
+  const initialSources = useAppSelector(state => state.lab.initialSources);
+  let startCode = 'print("Hello world!")';
+
+  if (initialSources?.source) {
+    const parsedSource = JSON.parse(initialSources.source);
+    startCode = parsedSource['main.py'];
+  }
 
   const handleRun = () => {
     const parsedData = data ? (data as PermissionResponse) : {permissions: []};
     // For now, restrict running python code to levelbuilders.
     if (parsedData.permissions.includes('levelbuilder')) {
       dispatch(appendOutput('Running code...'));
-      runPythonCode(code);
+      if (source) {
+        // TODO: will need to handle a potentially nested main.py
+        const code = source['main.py'];
+        if (typeof code === 'string') {
```

Comment:
What would this be if not a string?

Reply:
`main.py` could also theoretically be a folder name which would mean the contents are an object.